File: java/src/com/company/HeapSort.java
package com.company;

public class HeapSort {

    private static void heapify(Comparable[] a){
        int size = a.length;
        for(int i=(size/2)-1; i>=0; i--){
            sink(a,i,size);
        }
    }

    private static void sink(Comparable[] a,int index,int size){
        while(2*index+1<=size-1){
            int larger_child = 2*index+1;
            if(larger_child<size && less(a[larger_child],a[larger_child+1])){
                larger_child++;
            }
            if(less(a[larger_child],a[index])){
                break;
            }
            swap(a,index,larger_child);
            index = larger_child;
        }
    }

    private static boolean less(Comparable v,Comparable w){
        return v.compareTo(w)<0;
    }

    private static void show(Comparable[] a){
        for(int i = 0; i < a.length; i++){
            System.out.println(a[i]);
        }
    }

    private static boolean isSort(Comparable[] a){
        for (int i = 1; i < a.length; i++){
            if(less(a[i],a[i-1])){
                return false;
            }
        }
        return true;
    }

    private static void swap(Comparable[] a,int i,int j){
        Comparable temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }

    public static void sort(Comparable[] a){
        heapify(a);
        int size = a.length;
        int idx = size - 1;
        while(idx>=0){
            swap(a,0,idx--);
            sink(a,0,size);
        }
        show(a);
    }

    public static void main(String[] args) {
        String[] arr = new String[]{"T","G","S","H","R","I","P","E","N","A","O"};
        sort(arr);

    }
}

File: python/heapSort.py
class HeapSort:
    def __init__(self,arr):
        self.arr = arr
        self.index = len(arr)-1
    
    def __heapify(self):
        last_parent = (self.index-1)//2
        for i in range(last_parent,-1,-1):
            self.__sink(i,len(self.arr)-1)
    
    def __sink(self,idx,last_index):
        while((2*idx)+1 <= last_index):
            larger_child = (2*idx)+1
            if larger_child < last_index:
                if self.__isLess(larger_child,larger_child+1):
                    larger_child+=1
            if self.__isLess(larger_child,idx):
                break
            self.__swap(larger_child,idx)
            idx = larger_child
    
    def __swap(self,i,j):
        temp = self.arr[i]
        self.arr[i] = self.arr[j]
        self.arr[j] = temp

    def __isLess(self,i,j):
        return self.arr[i] < self.arr[j] 
    
    def sort(self):
        self.__heapify()
        while(self.index>=1):
            self.__swap(0,self.index)
            self.index-=1
            self.__sink(0,self.index)

    def showHeap(self):
        print(self.arr)

if __name__ == "__main__":
    arr = [31, 18, 20, 17, 18, 36, 44, 35, 22, 14]
    heap = HeapSort(arr)
    heap.sort()
    heap.showHeap()
